src/execution_plan/ops/op_aggregate.c

```diff
@@ -13,37 +13,35 @@
 #include "../../arithmetic/aggregate.h"
 
 /* Forward declarations. */
-static OpResult AggregateInit(OpBase *opBase);
 static Record AggregateConsume(OpBase *opBase);
 static OpResult AggregateReset(OpBase *opBase);
 static void AggregateFree(OpBase *opBase);
 
-/* Initialize expression_classification, which denotes whether each
- * expression in the RETURN or ORDER segment is an aggregate function.
- * In addition keeps track of non-aggregated expressions in
- * a separate array.*/
-static void _classify_expressions(OpAggregate *op) {
-	op->non_aggregated_expressions = array_new(AR_ExpNode *, 0);
+/* Migrate each expression projected by this operation to either
+ * the array of keys or the array of aggregate functions as appropriate. */
+static void _migrate_expressions(OpAggregate *op, AR_ExpNode **exps) {
+	uint exp_count = array_len(exps);
+	op->key_exps = array_new(AR_ExpNode *, 0);
+	op->aggregate_exps = array_new(AR_ExpNode *, 1);
 
-	op->expression_classification = rm_malloc(sizeof(ExpClassification) * op->exp_count);
-
-	for(uint i = 0; i < op->exp_count; i++) {
-		AR_ExpNode *exp = op->exps[i];
+	for(uint i = 0; i < exp_count; i++) {
+		AR_ExpNode *exp = exps[i];
 		if(!AR_EXP_ContainsAggregation(exp)) {
-			op->expression_classification[i] = NON_AGGREGATED;
-			op->non_aggregated_expressions = array_append(op->non_aggregated_expressions, exp);
+			op->key_exps = array_append(op->key_exps, exp);
 		} else {
-			op->expression_classification[i] = AGGREGATED;
+			op->aggregate_exps = array_append(op->aggregate_exps, exp);
 		}
 	}
+
+	op->aggregate_count = array_len(op->aggregate_exps);
+	op->key_count = array_len(op->key_exps);
 }
 
-static AR_ExpNode **_build_aggregated_expressions(OpAggregate *op) {
-	AR_ExpNode **agg_exps = array_new(AR_ExpNode *, 1);
+static AR_ExpNode **_build_aggregate_exps(OpAggregate *op) {
+	AR_ExpNode **agg_exps = array_new(AR_ExpNode *, op->aggregate_count);
```

Comment:
`AR_ExpNode **agg_exps = rm_malloc(sizeof(AR_ExpNode*) * op->aggregate_count);`
This will require a change in `group.c`
not mandatory, but we're not really using dynamic array capabilities.